### ISSUE TITLE
logrotate breaks log sync

### ISSUE THREAD

**Issue by ubellavance** [axkibe/lsyncd#282]
Hi,

I'm using lsyncd to sync log files form production servers to developper-accessible servers.  The developpers need to be able to open the sync'd log file and to a tail -f on it as well.  In my testing, I figured out that I would need to use --append so that the tail -f works. However, when the logrotate process on the source runs and rotates the logs, it looks like the sync process is broken.  Should I try --append-verify?

**Comment by axkibe** [comment]
I don't fully understand this, but I guess as far I get it, this isn't an Lsyncd issue.

**Comment by ubellavance** [comment]
I can't remember, it was 2.5 years ago, but I found a working solution, except for the hard links on [log rotation](https://groups.google.com/forum/#!topic/lsyncd/tXP4bndW5qY).

**Comment by axkibe** [comment]
Ya, hard links are a fundamental problem of event reporting.

I see no way around them otherwise as to suggest to use symlinks instead.

**Comment by ubellavance** [comment]
A solution would be to make sure rsync runs with -H when lsyncd is reloaded.  I'd simply run a reload of lsyncd just after the log rotation.

**Comment by axkibe** [comment]
-H doesn't cut it, since Lsyncd calls rsync with a custom filter passed via pipe, so rsync only has to look at files that changed. So neither Lsyncd nor rsync know of other hard links that point to the same node on the filesystem.

**Comment by ubellavance** [comment]
I thought that lsynd was doing a full rsync on restart (w/o filter).

**Comment by axkibe** [comment]
I does. But afterward hard linked files might run out of sync, because the event is only received for the path the file was opened with.

**Comment by ubellavance** [comment]
The hard link changes only once a week: Thursday, 00:00:00. So I tried running a restart of lsyncd on Thu, 04:10:00 but it didn't work. Would you want me to try again? Do you want me to create a new issue on github?

**Comment by axkibe** [comment]
The issue is the following. Suppose you have two files hardlinked to the same inode.

/src/fileA
/src/fileB

When an application now opens fileA and changes it, the Linux kernel reports to Lsyncd "fileA" in the directory you watched as "/src" has changed.

Neither the kernel nor Lsyncd know of fileB to be the same thing. Even the kernel only knows for "fileA" there are 2 file hardlinks pointing there, it doesn't know it is fileB.

Lsyncd then calls rsync, "rsync dude, /src/fileA has changed you ought to trasmit it". So rsync doesn't check fileB either.

**Comment by ubellavance** [comment]
I understand perfectly that, but as I said, my problem would be solved if lsyncd would actually do a full sync on a restart, which was not the case in my tests. Right now I use a cronjob to start a series (one per app) of rsync to make sure that the hard link is sync'd. If lsyncd would do the full sync correctly, I would only have to run an lsyncd restart just after the logs rotation. I'd be fine with that.

**Comment by axkibe** [comment]
Sorry, but it does make a full sync on restart.

**Comment by ubellavance** [comment]
Why sorry? I'll test it again. I'll let you know how it goes next Thursday. If it doesn't work, what kind of info would you like to have? Special config?

**Comment by axkibe** [comment]
The output of a minimal as possible configuration that stilll doesn't behave as expected with "-log Exec" turned out.

**Comment by ubellavance** [comment]
Ok. I'll start as is. If it doesn't work next Thursday, I'll configure the -log directive.

Here's my cronjob: `0 1 * * 4 /etc/init.d/lsyncd restart`  

Since the log rotation occurs at 00:00 Thursday, it should sync everyting 1 hour later. I'll let you know how it goes. I have disabled the rsync cronjobs.

**Comment by ubellavance** [comment]
I have the logs you requested.  Is there a way to send them somewhere private?

**Comment by ubellavance** [comment]
@axkibe It looks like it is the --append flag that I'm using that prevents the full startup sync to work.

```
sync {
    default.rsyncssh,
    source     = "/var/www/production/client1-app1/logs",
    host       = "dev.example.com",
    targetdir  = "/var/www/extlogs/production/client1-app1/",


    rsync      = {
       hard_links = true,
       _extra = {"--append"}
    }

}
```
If I remove the `_extra = {"--append"}`, the full startup sync works.  I am using the --append option because the files are web server logs and I wanted to make it possible to do a tail -f on the log file at the destination.  Would there be a way to tell lsyncd to drop the --append flag on the restart?
